### PR TITLE
fix: support macOS x86 PyTorch sync

### LINKED ISSUE #335: [Bug]: setting up cuga env is not working on my Mac
### What happened

error: Distribution `torch==2.11.0 @ registry+https://download.pytorch.org/whl/cpu` can't be installed because it doesn't have a source distribution or wheel for the current platform

hint: You're on macOS (`macosx_15_0_x86_64`), but `torch` (v2.11.0) only has wheels for the following platform: `macosx_11_0_arm64`; consider adding your platform to `tool.uv.required-environments` to ensure uv resolves to a version with compatible wheels

### How to reproduce

Apple M3 Max, Mac version 15.3, cuga agent is not getting installed.

### Environment

Latest version from https://github.com/cuga-project/cuga-agent.git 

### Logs, screenshots, or config

_No response_

### Checklist

- [x] I searched existing issues and this doesn’t look like a duplicate

### PR DESCRIPTION
**Pull request by sami-marreed**
## Summary
- Add macOS x86_64/Rosetta-specific PyTorch constraints so uv resolves the last compatible Intel Mac wheels.
- Regenerate `uv.lock` with separate macOS x86_64 torch/torchvision entries while preserving current Linux CPU resolution.
- Add dependency sync smoke coverage for macOS x86_64 and Docker/UBI-shaped Linux targets.

## Test plan
- `uv lock --check`
- `uv sync --locked`
- `uv run python -c "import torch, torchvision; print(torch.__version__, torchvision.__version__)"`
- `MACOSX_DEPLOYMENT_TARGET=15.0 uv sync --locked --dry-run --python-platform x86_64-apple-darwin`
- `uv sync --locked --dry-run --no-editable --no-dev --extra observability --python-platform x86_64-unknown-linux-gnu`
- `uv sync --locked --dry-run --no-editable --no-dev --extra observability --python-platform aarch64-unknown-linux-gnu`

Closes #335

## Summary by CodeRabbit

* **Chores**
  * Added cross-platform CI smoke checks to validate dependency synchronization across macOS (Intel) and Linux (amd64/arm64), including an expected failure case for incompatible lockfile scenarios.
  * Updated dependency requirements to be conditional by platform/architecture and Python version, with stricter PyTorch/Torchvision caps on Intel macOS to maintain compatibility.

### CODE REVIEW

File: uv.lock [hunk 1]
```diff
@@ -7472,21 +7548,43 @@ wheels = [
...
-    { name = "pillow", marker = "sys_platform == 'darwin'" },
-    { name = "torch", version = "2.11.0", source = { registry = "https://download.pytorch.org/whl/cpu" }, marker = "sys_platform == 'darwin'" },
+    { name = "numpy", version = "2.2.6", source = { registry = "https://pypi.org/simple" }, marker = "python_full_version < '3.11' and platform_machine != 'x86_64' and sys_platform == 'darwin'" },
+    { name = "numpy", version = "2.4.4", source = { registry = "https://pypi.org/simple" }, marker = "python_full_version >= '3.11' and platform_machine != 'x86_64' and sys_platform == 'darwin'" },
+    { name = "pillow", marker = "platform_machine != 'x86_64' and sys_platform == 'darwin'" },
+    { name = "torch", version = "2.11.0", source = { registry = "https://download.pytorch.org/whl/cpu" }, marker = "platform_machine != 'x86_64' and sys_platform == 'darwin'" },
 ]
 wheels = [
     { url = "https://download-r2.pytorch.org/whl/cpu/torchvision-0.26.0-cp310-cp310-macosx_11_0_arm64.whl", hash = "sha256:a06d4772a8e13e772906ed736cc53ec6639e5e60554f8e5fa6ca165aabebc464", upload-time = "2026-03-23T15:36:09Z" },
```
Comment (offerakrabi) [inline uv.lock:7582]: What is wrong: [uv.lock](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/index.html?id=5e14e769-6130-4a62-9217-ad40360d2e95&parentId=1&origin=fb43c0a7-c025-4989-9b6c-9f46e457e5f8&swVersion=5&extensionId=ZooCodeOrganization.zoo-code&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app&purpose=webviewView) includes macOS x86_64 Python 3.13 resolution markers for [torch](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/index.html?id=5e14e769-6130-4a62-9217-ad40360d2e95&parentId=1&origin=fb43c0a7-c025-4989-9b6c-9f46e457e5f8&swVersion=5&extensionId=ZooCodeOrganization.zoo-code&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app&purpose=webviewView), but the wheel list only contains cp310, cp311, and cp312 wheels at [uv.lock](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/index.html?id=5e14e769-6130-4a62-9217-ad40360d2e95&parentId=1&origin=fb43c0a7-c025-4989-9b6c-9f46e457e5f8&swVersion=5&extensionId=ZooCodeOrganization.zoo-code&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app&purpose=webviewView). A read-only dry-run with Python 3.13 failed with no compatible wheel/source distribution.

Why it matters: The project supports Python <3.14 in [uv.lock](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/index.html?id=5e14e769-6130-4a62-9217-ad40360d2e95&parentId=1&origin=fb43c0a7-c025-4989-9b6c-9f46e457e5f8&swVersion=5&extensionId=ZooCodeOrganization.zoo-code&platform=electron&vscode-resource-base-authority=vscode-resource.vscode-cdn.net&parentOrigin=vscode-file%3A%2F%2Fvscode-app&purpose=webviewView), so Intel Mac / Rosetta users on Python 3.13 still cannot sync dependencies.

How to fix: Either drop/enforce macOS x86_64 Python 3.13 support, or change the dependency markers/resolution so Python 3.13 does not select unsupported Intel Mac PyTorch wheels. Add a Python 3.13 macOS x86_64 dry-run CI lane.

File: .github/workflows/smoke-pip-install.yml [hunk 2]
```diff
@@ -12,6 +12,39 @@ concurrency:
...
+    name: Dependency sync (${{ matrix.name }})
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 25
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - name: macOS x86_64
+            os: ubuntu-latest
+            command: MACOSX_DEPLOYMENT_TARGET=15.0 uv sync --locked --dry-run --no-dev --python-platform x86_64-apple-darwin
```
Comment (offerakrabi) [inline .github/workflows/smoke-pip-install.yml]: What is wrong: The macOS x86_64 dry-run command does not specify Python 3.13, and the job sets up only Python 3.12 at [.github/workflows/smoke-pip-install.yml](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/.github/workflows/smoke-pip-install.yml:43).

Why it matters: CI misses the failing supported-platform combination that the lock currently advertises.

How to fix: Matrix the smoke job over Python 3.12 and 3.13, passing the selected interpreter explicitly to uv.

**Review by offerakrabi — COMMENTED** [review]
i did not find anything, so attaching the agent's report

### Strengths
- Correctly adds macOS x86_64-specific PyTorch constraints in [`pyproject.toml`](pyproject.toml:51).
- [`uv.lock`](uv.lock:7454) now has separate Intel macOS [`torch`](uv.lock:7454) / [`torchvision`](uv.lock:7552) entries and preserves Linux CPU resolution.
- CI adds low-cost dry-run resolver coverage in [`.github/workflows/smoke-pip-install.yml`](.github/workflows/smoke-pip-install.yml:15).

### Issues

One-line issue list:
- Critical — comment on [`uv.lock`](uv.lock:7458): macOS x86_64 Python 3.13 is locked, but [`torch`](uv.lock:7472) 2.2.2 has only cp310-cp312 Intel Mac wheels.
- Important — comment on [`.github/workflows/smoke-pip-install.yml`](.github/workflows/smoke-pip-install.yml:25): CI tests only Python 3.12 for macOS x86_64, missing the broken Python 3.13 path.
- Minor — comment on [`.github/workflows/smoke-pip-install.yml`](.github/workflows/smoke-pip-install.yml:23): job name implies macOS runtime coverage, but it is an Ubuntu-hosted resolver dry-run.

### Recommendations
- Add explicit Python-version coverage for macOS x86_64 resolver smoke tests.
- Either support macOS x86_64 Python 3.13 or document/enforce that Intel Mac PyTorch support ends at Python 3.12.
- Keep the dry-run strategy; it catches resolution failures without full dependency installation cost.

### Assessment
Ready to merge? No.
Reasoning: The PR fixes the Python 3.12 macOS x86_64 path, but the lock still claims Python 3.13 macOS x86_64 support and that dry-run fails because [`torch`](uv.lock:7454) 2.2.2 has no cp313 Intel Mac wheel.

Full issue list:

#### Critical (Must Fix)
- Comment line: [`uv.lock`](uv.lock:7458)
- What is wrong: [`uv.lock`](uv.lock:7458) includes macOS x86_64 Python 3.13 resolution markers for [`torch`](uv.lock:7454), but the wheel list only contains cp310, cp311, and cp312 wheels at [`uv.lock`](uv.lock:7473). A read-only dry-run with Python 3.13 failed with no compatible wheel/source distribution.
- Why it matters: The project supports Python <3.14 in [`uv.lock`](uv.lock:3), so Intel Mac / Rosetta users on Python 3.13 still cannot sync dependencies.
- How to fix: Either drop/enforce macOS x86_64 Python 3.13 support, or change the dependency markers/resolution so Python 3.13 does not select unsupported Intel Mac PyTorch wheels. Add a Python 3.13 macOS x86_64 dry-run CI lane.

#### Important (Should Fix)
- Comment line: [`.github/workflows/smoke-pip-install.yml`](.github/workflows/smoke-pip-install.yml:25)
- What is wrong: The macOS x86_64 dry-run command does not specify Python 3.13, and the job sets up only Python 3.12 at [`.github/workflows/smoke-pip-install.yml`](.github/workflows/smoke-pip-install.yml:43).
- Why it matters: CI misses the failing supported-platform combination that the lock currently advertises.
- How to fix: Matrix the smoke job over Python 3.12 and 3.13, passing the selected interpreter explicitly to uv.

#### Minor (Nice to Have)
- Comment line: [`.github/workflows/smoke-pip-install.yml`](.github/workflows/smoke-pip-install.yml:23)
- What is wrong: The job is named macOS x86_64, but it runs on Ubuntu and only cross-resolves dependencies.
- Why it matters: The label overstates coverage; it does not verify runtime import behavior on real macOS/Rosetta.
- How to fix: Rename it to macOS x86_64 resolver dry-run, or add a separate real macOS runtime smoke if needed.

File: uv.lock (hunk 1 above)
Comment (sami-marreed) [reply uv.lock:7582]: Fixed in 8821e6b7 — Intel Mac Py3.13 excluded via tool.uv.environments; torch/torchvision markers capped at python_version < '3.13'. Lock no longer claims cp313 Intel Mac torch wheels.

File: .github/workflows/smoke-pip-install.yml (hunk 2 above)
Comment (sami-marreed) [reply .github/workflows/smoke-pip-install.yml]: Fixed in 8821e6b7 — matrix includes Py3.12 pass lane and Py3.13 negative test (expects lockfile rejection). --python passed explicitly; job renamed to resolver dry-run.

[the remark this answers is not in the document]
Comment (sami-marreed) [reply .github/workflows/smoke-pip-install.yml]: Fixed in 8821e6b7 — added --extra observability and --no-editable to the macOS command.

**Comment by sami-marreed** [comment]
Updates addressing review (HEAD `8821e6b7`):

**Addressed**

- **#c1** — Fixed — Intel Mac Py3.13 excluded from universal lock via `tool.uv.environments`; torch/torchvision markers capped at `python_version < '3.13'`. Transitive deps (accelerate via docling) were still resolving torch 2.2.2 for Py3.13; lock regenerated with no cp313 Intel Mac torch entries (`pyproject.toml`, `uv.lock`)
- **#c2** — Fixed — CI matrix adds Py3.12 pass lane and Py3.13 negative test asserting lockfile rejection; `--python` passed explicitly on all rows (`.github/workflows/smoke-pip-install.yml`)
- **#c3** — Fixed — job renamed to `macOS x86_64 resolver dry-run` to reflect Ubuntu-hosted resolver check
- **#c4** — Fixed — macOS command now includes `--extra observability` and `--no-editable` for parity with Linux lanes
- **#c5** — Fixed — `permissions: contents: read` and `persist-credentials: false` on checkout

**Not in this pass**

- **#c6** — Deferred — action SHA pinning; rest of repo uses `@v5`/`@v6` tags. Can do repo-wide pin in a follow-up if desired.

**Platform note:** Intel Mac x86_64/Rosetta is supported on Python 3.10–3.12 only (PyTorch 2.2.2 wheel ceiling). Py3.13 on other platforms is unaffected.

Ready for another look.

**Review by offerakrabi — APPROVED** [review]

**Comment by sami-marreed** [comment]
@offerakrabi i think this is dangerous pr what u think?

**Comment by offerakrabi** [comment]
> @offerakrabi i think this is dangerous pr what u think?

I think the danger is introducing dependency clashes.

I must say, i don't understand the requirement. the intel macs are at least 6 years old..
if it works on windows/linux/mac it is good enough for me
